ctapipe/tools/stage1.py

```diff
@@ -0,0 +1,767 @@
+            append=True,
+            serialize_meta=serialize_meta,
+        )
+        for telescope_type in subarray.telescope_types:
+            ids = set(subarray.get_tel_ids_for_type(telescope_type))
+            if len(ids) > 0:  # only write if there is a telescope with this camera
+                tel_id = list(ids)[0]
+                camera = subarray.tel[tel_id].camera
+                camera.geometry.to_table().write(
+                    self.output_filename,
+                    path=f"/configuration/instrument/telescope/camera/geometry_{camera}",
+                    append=True,
+                    serialize_meta=serialize_meta,
+                )
+                camera.readout.to_table().write(
+                    self.output_filename,
+                    path=f"/configuration/instrument/telescope/camera/readout_{camera}",
+                    append=True,
+                    serialize_meta=serialize_meta,
+                )
+
+    def _write_processing_statistics(self):
+        """ write out the event selection stats, etc. """
+        image_stats = self.check_image.to_table(functions=True)
+        image_stats.write(
+            self.output_filename,
+            path="/dl1/service/image_statistics",
+            append=True,
+            serialize_meta=True,
+        )
+
+    def _parameterize_image(self, subarray, data, tel_id):
+        """Apply Image Cleaning
+       
+        Parameters
+        ----------
+        subarray : SubarrayDescription
+           subarray description
+        data : DL1CameraContainer
+            calibrated camera data
+        tel_id: int
+            which telescope is being cleaned
+        
+        Returns
+        -------
+        np.ndarray, ImageParametersContainer: 
+            cleaning mask, parameters
+        """
+
+        tel = subarray.tel[tel_id]
+
+        # apply cleaning
+
+        mask = self.clean(tel_id=tel_id, image=data.image)
+
+        clean_image = data.image.copy()
+        clean_image[~mask] = 0
+
+        params = ExtendedImageParametersContainer()
+
+        # check if image can be parameterized:
+
+        image_criteria = self.check_image(clean_image)
+        self.log.debug(
+            "image_criteria: %s",
+            list(zip(self.check_image.criteria_names[1:], image_criteria)),
+        )
+
+        # parameterize the event if all criteria pass:
+        if all(image_criteria):
+            params.hillas = hillas_parameters(
+                geom=tel.camera.geometry, image=clean_image
+            )
+            params.timing = timing_parameters(
+                geom=tel.camera.geometry,
+                image=clean_image,
+                pulse_time=data.pulse_time,
+                hillas_parameters=params.hillas,
+            )
+            params.leakage = leakage(
+                geom=tel.camera.geometry, image=data.image, cleaning_mask=mask
+            )
+            params.concentration = concentration(
```

Comment:
I would suggest to only pass in `geom[mask]` and `image[mask]` here. concentration is quite slow if you do this for all pixels. (needs to rotate them into the shower coordinates and sort by brightness)